Add friends likes

rovniipoc opened the pull request:
Здравствуйте.
Представляю на ревью проект по ТЗ спринта 11.

rovniipoc committed:
Доработать по результатам 2-го ревью

Изменения:
Исправлено логирование методов findAll()

rovniipoc committed:
Merge remote-tracking branch 'origin/main'

rovniipoc committed:
Добавить FilmStorage и UserStorage

rovniipoc committed:
Добавить UserService и FilmService

rovniipoc committed:
Реализовать ТЗ спринта 11

Изменения:
- добавлен ErrorHandler и ErrorResponse
- добавлены новые поля в User и Film
- доработаны (расширен функционал) контроллеров, сервисов, хранилищ

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+//    @ExceptionHandler
+//    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
```
Это верно, раскомментрируй

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,73 @@
-    private static final LocalDate VALID_RELEASE_DATE = LocalDate.of(1895, 12, 28);
-    private final Map<Long, Film> films = new HashMap<>();
-    private Long idCounter = 0L;
+    private final FilmStorage filmStorage;
```
контроллеры обращаются только в сервиса, в сторадж они не должны ходить, даже если в сервиса будет просто проксирование запроса
А уже сервис ходит или двругие сервиса или в хранилища

YazykovMikhail replied:
Аналогично для User

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,73 @@
 
 @RestController
 @RequestMapping("/films")
 @Slf4j
```
Стандартный шаблон для логов
createMethod(Request request) {
log.info("пришел Post запрос /films с телом: {}", request)
Response response = //логика метода, если есть бизнес процессы для логирвоания логируем их
log.info("Отправлен ответ Post /films с телом: {}", response)
}
что бы было видно что пришло, что ушло и если были какие то изменения в процессы выполнянения тоже были видны, потому что ты пойдешь в логи когда у тебя что то сломалось, просто по сообщения "Фильм добалвен" ничего не понятно.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -17,5 +19,6 @@ public class Film {
     private LocalDate releaseDate;
     @Positive
     private Long duration;
+    private Set<Long> likes = new HashSet<>();
```
Используй @JsonIgnore
ЧТо бы через эндпоинты для Film нельзя было влиять на лайки, ведь для них есть отделные эндпоинты а тут они просто храняться

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -4,17 +4,19 @@
     private String name;
     @Past
     private LocalDate birthday;
+    private Set<Long> friends = new HashSet<>();
```
Используй @JsonIgnore
ЧТо бы через эндпоинты для User нельзя было влиять на друзей, ведь для них есть отделные эндпоинты а тут они просто храняться

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,76 @@
+        return films.get(id);
+    }
+
+    private void releaseDateValidation(Film film) throws ValidationException {
```
Эту проверку вынеси в сервис, нет смысла вриходит в харанилище, если каких-то бизнес данных не хватает, а именно сервис отвечает за бизнес процессы

rovniipoc committed:
Устранить замечания первого ревью 11 спринта

Изменения:
- добавлен обработчик исключений MethodArgumentNotValidException и Throwable;
- контроллеры теперь общаются только с сервисами (нет прямого доступа в хранилища);
- логирование контроллеров сделано по шаблону;
- добавлены аннотации @JsonIgnore к полям likes и friends;
- необходимые проверки, относящиеся к бизнес-логике вынесены из хранилища в сервисы
